### PR TITLE
Make hardcoded ROLE_SONATA_ADMIN and ROLE_SUPER_ADMIN configurable

### LINKED ISSUE #4290: Make ROLE_SONATA_ADMIN (hardcoded in twig templates) configurable
In [`add_block.html.twig`](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/Resources/views/Core/add_block.html.twig#L22) and [`standard_layout.html.twig`](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/Resources/views/standard_layout.html.twig#L187) the role `ROLE_SONATA_ADMIN` is hardcoded. What about making it configurable?

This would allow us to configure a custom role i.e. the standard `ROLE_ADMIN` role, dropping the useless `ROLE_SONATA_ADMIN` role. Of course the default value would be the `ROLE_SONATA_ADMIN` for BC.

Example:

```yml
sonata_admin:
    # ...
    security:
        admin_role: ROLE_ADMIN
        admin_permission: # ...
```

### PR DESCRIPTION
**Pull request by gremo**
I am targeting this branch, because this is an enhancement.

Closes #4290

## Changelog
```markdown
### Added
- Ability to configure sonata main admin and super admin roles under the `security` configuration key
```

## Subject
As #4301 seems abandoned, I'm proposing this PR, which should be backward compatible.

Sonata main roles (`ROLE_SONATA_ADMIN` and `ROLE_SUPER_ADMIN`) used in Twig templates and in the security middleware are now configurable under `security` configuration key. This means you are not forced anymore to build your role hierarchy based on the above roles but you can choose your own (i.e. stick with the "standard" roles `ROLE_ADMIN` and `ROLE_SUPER_ADMIN`).

Example configuration:

```yml
sonata_admin:
    # ...
    security:
        # ...
        role_admin: ROLE_ADMIN
        role_super_admin: ROLE_SUPER_ADMIN
```
## Relation to other sonata bundles
- SonataUserBundle: changes needed https://github.com/sonata-project/SonataUserBundle/pull/909
- SonataMediaBundle: changes needed https://github.com/sonata-project/SonataMediaBundle/pull/1297

### CODE REVIEW

File: DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -125,6 +125,14 @@ public function getConfigTreeBuilder()
                             ->defaultTrue()
                             ->info('Enable automatic registration of annotations with JMSDiExtraBundle')
                         ->end()
+                        ->scalarNode('role_admin')
+                            ->defaultValue('ROLE_SONATA_ADMIN')
+                            ->cannotBeEmpty()
```
Comment (greg0ire) [inline DependencyInjection/Configuration.php]: Can you use `info` here, to explain what granting this role does?

Comment (gremo) [inline DependencyInjection/Configuration.php]: Sure, I'm at work now but I'll look at it this evening. What about moving the configuration under the security key?

Comment (greg0ire) [inline DependencyInjection/Configuration.php]: Would be great, but you would have to do it in a BC way

Comment (jordisala1991) [inline DependencyInjection/Configuration.php]: it will be bc anyways because it is a new key, or am I missing something @greg0ire ?

Comment (greg0ire) [inline DependencyInjection/Configuration.php]: @gremo said move, not copy ;)

Comment (gremo) [inline DependencyInjection/Configuration.php]: Can't get it: why adding keys with default values (without altering how the bundle works) it's a BC break?

Comment (greg0ire) [inline DependencyInjection/Configuration.php]: Oh right, they weren't already there... Sorry, I'm tired.

**Comment by gremo** [comment]
@greg0ire forced push. Moved both keys under `security` for consistency and added infos.

File: DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -63,6 +63,16 @@ public function getConfigTreeBuilder()
                             ->defaultValue(array('CREATE', 'LIST', 'DELETE', 'UNDELETE', 'EXPORT', 'OPERATOR', 'MASTER'))
                             ->prototype('scalar')->end()
                         ->end()
+                        ->scalarNode('role_admin')
+                            ->cannotBeEmpty()
+                            ->defaultValue('ROLE_SONATA_ADMIN')
+                            ->info('Role which will see the top nav bar and dropdown groups regardless its configuration')
```
Comment (greg0ire) [inline DependencyInjection/Configuration.php]: "regardless of its configuration"

File: DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -63,6 +63,16 @@ public function getConfigTreeBuilder()
...
+                        ->end()
+                            ->scalarNode('role_super_admin')
+                            ->cannotBeEmpty()
+                            ->defaultValue('ROLE_SUPER_ADMIN')
+                        ->info('Role which will perform all admin actions, see dashboard, menu and search groups regardless its configuration')
```
Comment (greg0ire) [inline DependencyInjection/Configuration.php]: Same here

**Comment by gremo** [comment]
@greg0ire Done! Oh my... English :)

**Comment by greg0ire** [comment]
We need you to write a bit more of it :P in the docs please

**Comment by gremo** [comment]
@greg0ire can you please point me to the file I need to edit?

**Comment by greg0ire** [comment]
@gremo I have no idea, I don't know the doc more than you do, but maybe you can grep `ROLE_` in it?

**Comment by OskarStark** [comment]
@gremo it would be great, if you could finalize this PR

**Comment by gremo** [comment]
I updated the documentation and rebased this PR, @greg0ire @OskarStark can you please check if it's OK now?
I mean, the documentation is very hard to read and understand, it's full of errors and should be written from scratch... I did my best 👍

**Comment by gremo** [comment]
@greg0ire can you please merge so I can delete the branch and cleanup my things? Thanks!

**Comment by greg0ire** [comment]
I cannot click the button because the tests are not passing, please fix them.

**Comment by gremo** [comment]
@greg0ire Ok, fixed I hope (blank like in rst file).

**Comment by greg0ire** [comment]
@sonata-project/contributors please review

**Comment by gremo** [comment]
@jlamur if I understand, you want me to edit CHANGELOG? Which version should I target?

**Comment by greg0ire** [comment]
Don't change any file please, just edit [this message](instead) (have a look at the PR template if you have some doubts)

**Comment by gremo** [comment]
@greg0ire please check, thanks.

**Comment by greg0ire** [comment]
Please write a changelog directed at end users. Keep in mind that they probably do not care what file or class you changed, but what features you provided / what bugs you fixed. Try to answer the question "What benefit do I get from upgrading?"
More on http://keepachangelog.com

**Comment by gremo** [comment]
@greg0ire changelog updated, please merge.

**Comment by greg0ire** [comment]
? You didn't change the changelog, did you?

**Comment by gremo** [comment]
@greg0ire accidentally delete, I've messed up something. Too late to do this stuff 😭

**Comment by greg0ire** [comment]
You're almost there :P

**Comment by gremo** [comment]
@greg0ire please check before I do something else wrong... 😁

**Comment by greg0ire** [comment]
See that: 
```
### Added
- Added `role_admin `and `role_super_admin` configuration roles under `security` key

### Changed
- `DependencyInjection/SonataAdminExtension.php` to pass these options to the admin pool
- `Resources/config/security.xml` in order to inject these options in the relevant services
- Twig templates in order to use the admin pool and retrieve the configuration roles
```

All you need to do is to make it a bit less about files and classes, and a bit more about functionality for the end user. Imagine I'm an end user. What will I get if this gets merged?

**Comment by gremo** [comment]
Hi found that there are some hardcoded roles in other bundles like SonataUserBundle. This means we have to unmerge this because changing the roles in the configuration (subject of this PR) will break other bundles.

For example: https://github.com/sonata-project/SonataUserBundle/blob/3.x/Security/EditableRolesBuilder.php#L110

What do you think?

**Comment by greg0ire** [comment]
> This means we have to unmerge this because changing the roles in the configuration (subject of this PR) will break other bundles.

... it's not merged yet, is it?

**Comment by gremo** [comment]
@greg0ire yes you are right, my bad, isn't merged yet.

**Comment by greg0ire** [comment]
:sweat_smile:

**Comment by Soullivaneuh** [comment]
I updated your changelog for a simpler one.

**Comment by OskarStark** [comment]
@jlamur please review again

**Comment by jlamur** [comment]
> Hi found that there are some hardcoded roles in other bundles like SonataUserBundle. This means we have to unmerge think about this because changing the roles in the configuration (subject of this PR) will break other bundles.
> 
> For example: https://github.com/sonata-project/SonataUserBundle/blob/3.x/Security/EditableRolesBuilder.php#L110
> 
> What do you think?

This is important! @core23 @greg0ire any idea on how to properly support that?

**Comment by core23** [comment]
Where's the problem? You could use the property as well in other bundles.

**Comment by gremo** [comment]
I'm working on a PR for user and media bundles too.

**Comment by jlamur** [comment]
> Where's the problem? You could use the property as well in other bundles.

If an user update the admin bundle but not the orm bundle, then the configuration key won't work in every cases (like this [one](https://github.com/sonata-project/SonataUserBundle/blob/3.x/Security/EditableRolesBuilder.php#L110), obviously).
Anyway, this should be merged and released **after** all other bundles are fixed and released.

**Comment by gremo** [comment]
> Anyway, this should be merged and released after all other bundles are fixed and released.

@jlamur not the case of SonataUserBundle, but I need to check other bundles too.

[Here](https://github.com/sonata-project/SonataUserBundle/blob/3.x/Security/EditableRolesBuilder.php#L110) using the admin pool `getOption` (providing the default if not found) will not break anything:

       $isMaster = $this->authorizationChecker->isGranted(
            $this->pool->getOption('role_super_admin', 'ROLE_SUPER_ADMIN')
        );

**Comment by jlamur** [comment]
@gremo Yes you're right. But still, if the modification you shown is released after this PR, then the feature is broken. And IMO it's kind of critical because it touches the security.

**Comment by greg0ire** [comment]
Maybe we should even add something to the `conflicts` section in case people try using the new admin bundle with an outdated user bundle

**Comment by jlamur** [comment]
Moreover, if we add a `conflicts` to the user bundle composer too (after this PR gets merged and released), we can remove the default value in the `getOption` call:

```diff
 $isMaster = $this->authorizationChecker->isGranted(
-    $this->pool->getOption('role_super_admin', 'ROLE_SUPER_ADMIN')
+    $this->pool->getOption('role_super_admin')
 );
```

**Comment by gremo** [comment]
> Moreover, if we add a conflicts to the user bundle composer too 

Why forcing people that wants to upgrade only SonataUserBundle for some reason, to upgrade SonataAdminBundle too? I'll leave the default value and only add conflict in the SonataAdminBundle. Let remove that default in 4.x...

**Comment by greg0ire** [comment]
> Moreover, if we add a conflicts to the user bundle composer too (after this PR gets merged and released), we can remove the default value in the getOption call:

user requires admin, so you could simply bump the requirement to achieve this

**Comment by gremo** [comment]
@OskarStark added two tests.

File: DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -63,6 +63,16 @@ public function getConfigTreeBuilder()
...
                         ->end()
+                        ->scalarNode('role_admin')
+                            ->cannotBeEmpty()
+                            ->defaultValue('ROLE_SONATA_ADMIN')
+                            ->info('Role which will see the top nav bar and dropdown groups regardless of its configuration')
+                        ->end()
+                            ->scalarNode('role_super_admin')
+                            ->cannotBeEmpty()
+                            ->defaultValue('ROLE_SUPER_ADMIN')
+                        ->info('Role which will perform all admin actions, see dashboard, menu and search groups regardless of its configuration')
```
Comment (core23) [inline DependencyInjection/Configuration.php]: Indention ist wrong

Comment (gremo) [reply DependencyInjection/Configuration.php]: I used the style of the original file, like https://github.com/gremo/SonataAdminBundle/blob/b12ab3fed7f81870e67bce3945393b4f9a1b03dc/DependencyInjection/Configuration.php#L128... where is wrong btw?

Comment (core23) [reply DependencyInjection/Configuration.php]: There is one tab missing in front of `->info...`

**Comment by gremo** [comment]
As both SonataMedia and SonataUser has been updated, what about finalizing this PR?

**Comment by greg0ire** [comment]
@jlamur please review

**Comment by gremo** [comment]
OMG I rebased it... you dropped the support for php <7!!! I need this and, after 3 months of waiting, I will not able to use it because I can switch to php 7.

Any way to apply this PR to a commit _before_ the commit that dropped the support?

**Comment by Soullivaneuh** [comment]
@gremo Short answer: no.

Version dropping **is** backward compatible according to semver and was voter here to be sure of the decision: https://github.com/sonata-project/dev-kit/issues/285

We can't handle all OS constraint. Plus, the migration from PHP 7 to PHP 7.1 should not be that hard.

Waiting that, you can still ignore platform req with composer, but we don't recommend that.

See also: https://sonata-project.org/blog/2017/10/9/workflow-change

You would fairly understand that we can't wait to have all PR merged, especially on this very active bundle, to make such modification.

This one done for migration. At the future, version dropping PR must be merged at very last time. @sonata-project/contributors are already warned about that: https://github.com/sonata-project/SonataAdminBundle/blob/3.x/CONTRIBUTING.md#dependencies-version-dropping

Regards.

**Comment by Soullivaneuh** [comment]
@gremo please read and vote: https://github.com/sonata-project/dev-kit/issues/315

**Comment by gremo** [comment]
@Soullivaneuh Yes, you're perfectly right. In my specific case the company for which I work for has a private server locked at PHP 5.5. It won't to be so easy to switch to PHP 7 as upgrading from PHP 7 to PHP 7.1 😄 

Voted 🙈  but... does this new vote means that there is a chance to revert the dropping for php 5.6?

I'll update this PR as soon as possible.

**Comment by greg0ire** [comment]
We are discussing it, but we are probably revert to supporting 5.6 again on stable branches.

**Comment by gremo** [comment]
Well this PR could be merged despite the fact to revert to PHP 5.6, right?

**Comment by greg0ire** [comment]
@soullivaneuh should we decide to merge this anyway in no one has tested it after one week?

**Comment by jlamur** [comment]
@gremo @OskarStark I already wrote my review ;) https://github.com/sonata-project/SonataAdminBundle/pull/4548#pullrequestreview-71499811

**Comment by greg0ire** [comment]
@jlamur please approve and merge, we can't.

**Comment by jlamur** [comment]
Thank you @gremo :tada: 👍

**Pull request merged**